### PR TITLE
Epic 7.1 Added Various Loaders for Quizzify (AI Avengers)

### PR DESCRIPTION
**Pull request by Abhinav-Telukunta**
# Pull Request - File Loaders Extension for Quizzify

## Summary
Added loaders for the following file types:
1. CSV
2. DOCX
3. PPTX
4. Youtube
5. WebPage
6. TXT
7. PDF

Added filtering as well. For example for csv, user can mention startRow and endRow and specific columns and based on that content, quiz questions would be generated.

## Changes
- Added above listed file loaders and thoroughly tested it.
- Created a file called as loaders.py and added all the loaders in it.
- Refactored URL Loader a bit and tweaked some parameters of RAG pipeline and added schemas for each file type in tool_registry to support filtering.
- Made filetype parameter as mandatory and it needs to be passed in request body along with url mandatory field.

## Testing
Each loader was tested using Adhoc Testing.

## Results
All six loaders are working perfectly fine with and without filtering criteria.

## Notes
- This feature extends the support for other loaders as well apart from pdf files which enhances file type support.

### YouTube Sample Request and Response
sample_request_youtube
sample_response_youtube

### Webpage Sample Request and Response
sample_request_webpage
sample_response_webpage

### TXT Sample Request and Response
sample_request_txt
sample_response_txt

### DOCX Sample Request and Response
sample_request_docx
sample_response_docx

### PPTX Sample Request and Response
sample_request_pptx
sample_response_pptx

### CSV Sample Request and Response
sample_request_csv
sample_response_csv

### PDF Sample Request and Response
sample_request_pdf
sample_response_pdf

## How to Test
1. Clone the repo in your local

2. Create and activate virtual environment

3. Use pip install -r requirements.txt to install required libraries.

4. Create .env file with ENV_TYPE, GCP_PROJECT_ID  and GOOGLE_API_KEY fields. Env type is dev, gcp_project_id is your project id from cloud console project, google_api_key is your api key from AI studio.

5. Then type ./local-start.sh to start the application.

6. Add sample requests and responses for each file type as mentioned above in screenshots and test it.

### CODE REVIEW

File: app/api/tool_utilities.py [hunk 1]
```diff
@@ -18,7 +18,10 @@ def load_config():
 
 def get_executor_by_name(module_path):
     try:
-        module = __import__(module_path, fromlist=['executor'])
+        try:
```
Comment (mikhailocampo) [inline app/api/tool_utilities.py:21]: Try to consolidate

File: app/api/tool_utilities.py [hunk 2]
```diff
@@ -82,11 +85,13 @@ def validate_file_input(input_name: str, input_value: Any):
             logger.error(error_message)
             raise InputValidationError(error_message)
         try:
-            ToolFile.model_validate(file_obj, from_attributes=True)  # This will raise a validation error if the structure is incorrect
-        except ValidationError:
-            error_message = f"Each item in the input `{input_name}` must be a valid ToolFile where a URL is provided"
-            logger.error(error_message)
-            raise InputValidationError(error_message)
+            if 'filetype' not in file_obj:
```
Comment (mikhailocampo) [inline app/api/tool_utilities.py:88]: Looks good, good work

File: app/services/tool_registry.py [hunk 3]
```diff
@@ -20,3 +20,31 @@ class ToolFile(BaseModel):
     filePath: Optional[str] = None
     url: str
     filename: Optional[str] = None
+    filetype: str
+
```
Comment (mikhailocampo) [inline app/services/tool_registry.py:24]: LGTM

**Review by mikhailocampo — COMMENTED** [review]
Change the target branch from `main` to `STAGING`

File: app/services/tool_registry.py (hunk 3 above)
Comment (mikhailocampo) [reply app/services/tool_registry.py:24]: This is great work, can this team also create documentation on the new loaders that they had created such as detailed information on the filters? Would love to see it so we can pass it to the front end team.

File: app/api/tool_utilities.py (hunk 2 above)
Comment (Abhinav-Telukunta) [reply app/api/tool_utilities.py:88]: thankyou!

File: app/services/tool_registry.py (hunk 3 above)
Comment (Abhinav-Telukunta) [reply app/services/tool_registry.py:24]: Sure @mikhailocampo . We will create documentation on usage of the loaders and their parameters, processing and responses.

File: app/api/tool_utilities.py (hunk 1 above)
Comment (Abhinav-Telukunta) [reply app/api/tool_utilities.py]: changed it.

**Comment by AaronSosaRamos** [comment]
Closed due inactivity

**Pull request closed without merging**